### PR TITLE
Collecting changes from PRs, adding Flowspec & Encoding support

### PR DESCRIPTION
**Pull request by thepacketgeek**
This PR adds the following features:
(including changes from other PRs and supersedes: #23 , #26 , #20, #21, #31) 
- Message encoding
- Roundtrip pcap testing
- SAFI Enums
- Flowspec support
- Additional public attributes on some message types (Open, update)
- From/Into for `Prefix`

### CODE REVIEW

**Comment by DevQps** [comment]
Good work! I hope to take a look somewhere end this week! Thanks a lot!

**Comment by DevQps** [comment]
I see that we still get an error on `bgp-fuzz`:

```
error[E0063]: missing fields `ADD_PATH_SUPPORT`, `BGPSEC_SUPPORT`, `ENHANCED_ROUTE_REFRESH_SUPPORT` and 7 other fields in initializer of `bgp_rs::Capabilities`

  --> fuzz_targets/msg_target.rs:14:23

   |

14 |         capabilities: Capabilities {

   |                       ^^^^^^^^^^^^ missing `ADD_PATH_SUPPORT`, `BGPSEC_SUPPORT`, `ENHANCED_ROUTE_REFRESH_SUPPORT` and 7 other fields

error: aborting due to previous error
```
Job: https://travis-ci.com/github/DevQps/bgp-rs/jobs/325190046

I will look at everything either this evening or tomorrow btw!

**Comment by DevQps** [comment]
I tried to grasp what was going wrong in job 47.1 but I am a bit clueless. It seems that somehow some kind of input made it crash but I cannot really understand what it is from the printed lines. I also see no way to download the artifacts in Travis-CI so we can inspect the dumped file sadly.. If you have a good plan on how we can fix this? If there is not easy way of figuring this out, we could also temporary remove the fuzzing part from this PR and try to merge it later. I will review the code now!

File: .gitignore
Comment (DevQps) [inline .gitignore]: I guess the standard best practice is to put these inside the user or global .gitignore since they then apply to all the projects that person works on. So I guess it would be better to remove `.DS_Store` and move it to a global or user .gitignore!

https://git-scm.com/docs/gitignore

> Patterns which a user wants Git to ignore in all situations (e.g., backup or temporary files generated by the user’s editor of choice) generally go into a file specified by core.excludesFile in the user’s ~/.gitconfig. Its default value is $XDG_CONFIG_HOME/git/ignore. If $XDG_CONFIG_HOME is either not set or empty, $HOME/.config/git/ignore is used instead.

File: .travis.yml
Comment (DevQps) [inline .travis.yml:30]: I was wondering: If I interpret this correctly it states that we only want to fuzz on stable right?
Is there a reason why we want to fuzz on stable but not others?

File: src/update/attributes.rs
Comment (DevQps) [inline src/update/attributes.rs:321]: That's very interesting haha :) Let's create an issue of this after we merge the PR.

**Review by DevQps — CHANGES_REQUESTED** [review]
I put it on request changes for now, but feel free to discuss anything you do not agree with. A lot of stuff has been added I see! I probably need to dive deeper in some RFCs again in the coming weeks as well, but everything looks great so far! Some issues we can create after we merge!:

- If possible we can try to discover GitHub Actions instead of Travis-CI
- Try to put the encoding part behind a feature flag (encoding is basically putting a constructed BGP object back to its binary form again right?). 
- Try to put flowspec behind a feature flag. (My understanding is that this is separate from encoding right? I should really read this RFC)
- I think we should also include in either the Actions or Travis CI pipelines some checks that we always enforce rust-fmt, no clippy warnings and docs for all public items. I think I already have something like that ready.

Thanks a lot for everything in advance! Let me know what you think 👍

**Comment by thepacketgeek** [comment]
Thanks for the review! I can recreate the fuzz crash locally with the `travis_fuzz.sh` script but also have no idea how to interpret the output. I'll do some research on this, as I think it would be nice to keep fuzzing.

**Comment by DevQps** [comment]
Thanks for that! I will research GitHub Actions in meanwhile and try to reproduce the current pipeline.

**Comment by thepacketgeek** [comment]
I was able to inspect the crash (using crate hfuzz run-debug) and realized there was a parsing error in the Notification header. Fuzzing for the win, thanks @TheBlueMatt !!

File: .travis.yml
Comment (thepacketgeek) [reply .travis.yml:30]: Yeah, this seems to be checking that the default rustup target is 'stable'. We could likely target multiple versions with the change to GithubActions?

**Comment by DevQps** [comment]
Yeah good plan! Let's do it after we migrate to GitHub Actions then! Also nice that we found another bug!

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by thepacketgeek** [comment]
Interesting on the code coverage, I would think the roundtrip parsing/encoding would exercise a lot more. I'll add some more unittesting to this PR soon

**Comment by DevQps** [comment]
In meanwhile I read the complete docs of GitHub actions, so I expect that I can write a pipeline somewhere this week / weekend hopefully!

**Comment by thepacketgeek** [comment]
@DevQps I appreciate your patience with this PR. I'm having trouble running kcov locally to test coverage, will hopefully figure it out soon

**Comment by DevQps** [comment]
No problem! Next week I have 3 days off so I will create an Action pipeline that we can hopefully use then as well!

I might also create some issues for other things we can improve. On the mrt-rs project someone requested to be able to parse path attributes directly in order to parse TABLE_DUMP_V2 attributes for example!

EDIT: Sorry for my delay as well.. Things sometimes seem busier then expected :)

**Comment by thepacketgeek** [comment]
@DevQps I finally got kcov working locally and confirmed this is just a targeting issue (coverage is around 95% for bgp-rs lib). I'm attempting a fix for travis to see if I can reproduce the correct results that I have locally

**Comment by thepacketgeek** [comment]
I'm not sure what's going on with the coverage. I've attached what I get locally.

Screenshot 2020-05-13 11 21 12


And there is now a `gimli` dependency in `pcap` -> `error-chain` that doesn't work with rustc 1.34. `pcap` hasn't changed so we can't pin a version, it would need to be pinned in `pcap` or `error-chain`

**Comment by DevQps** [comment]
Good work at least! Nice to see you working on it! It seems like a hassle to support all dependencies for 1.34. One of us could create e PR on pcap or error-chain to fix this. If you think this is a good way forward, let me know! I can create a PR at one of those I think. At the other side: If this is a type of problem that occurs more frequently we could also decide to drop support for Rust 1.34 and require users to upgrade. I actually wonder what reasons could be that someone would not use the latest compiler. The only reason I can think of is that their distro might not yet have packaged these versions.

I also do not always understand the coverage percentages either... If you zoom in on the coverage on one of these files. Do the results make sense individually or does the coverage that it generates not really make sense?

I hope to finish an Action pipeline on `mrt-rs` today! If you think that general pipeline is nice as well, we could also use it for this repository.

**Comment by thepacketgeek** [comment]
### 1.34 support:
  - @theBlueMatt would it work for you to pin to bgp-rs v 0.5 (current version) for where you need 1.34 support? I can only imagine keeping support for old versions to be more difficult as dependencies update and change.

### Coverage
  - I suspect there may be something with the differential support and existing state for codecov. When I run locally, coverage looks good (I will increase the coverage for the yellow files in the image above).
  - The [coverages here](https://codecov.io/gh/DevQps/bgp-rs/pull/27/diff) match what codecov is displaying, but not what I'm seeing locally. I'm going to try a couple more things

**Comment by TheBlueMatt** [comment]
I can always use my branch instead of the new changes upstream, though that’s somewhat of a shame. 1.34.2 isn’t really all that old, and it kinda sucks that various dependencies require something newer. My response to such issues is to always rip out the dependency (which is why I tried bgp-rs in the first place - no dependencies), but that’s how I’ve ended up solely with crates with no dependencies lol. In any case, I’m not offering to do the work, so my option here carries little weight.

**Comment by thepacketgeek** [comment]
@TheBlueMatt The crate builds fine for 1.34.2, it's the testing that breaks. I tried pinning to old versions of gimli, smallvec, arrayvec, etc. and something (maybe cargo) is still trying to use the newer versions.

**Comment by TheBlueMatt** [comment]
Oh, right, testing is not an issue (though it would be nice to do a “cargo build” on rust 1.34 on Travis) - cargo never runs tests for dependencies so dev-dependencies is really just there for the benefit of bgp-rs.

**Comment by DevQps** [comment]
Looks good! I see all checks have passed as well. Do you think we can merge this? Or are there still some open points that we should fix before we can do that? I managed to create a Action pipeline on one of my other repo's, it tests all combinations of nightly, stable, beta and the supported platforms. We could also add a scheduled fuzzer in there as well or/and add it to the standard push pipeline

**Comment by thepacketgeek** [comment]
yeah, I think so! Targeting the coverage correctly is something we can hopefully do with the Action (I've had better results with `cargo-kcov`, but it wasn't working well with the existing state in codecov.io).

**Comment by DevQps** [comment]
Alright I have merged it! I will close the remaining PRs and get started on the GitHub Action pipeline. I'll let you know when I have something! If you're still up for some work maybe you could get started with isolating the Flowspec encoding behind a feature request? If you want we could also meet on Discord again to discuss the way forward?

**Pull request merged**
